File: day01.py
from time import time


def part1(changes):
    return sum(changes)


def part2(changes):
    frequencies = set()
    result = 0
    while True:
        for change in changes:
            result += change
            if result in frequencies:
                return result
            else:
                frequencies.add(result)


if __name__ == '__main__':
    changes = []
    with open('inputs/01') as f:
        changes = list(map(int, filter(None, f.readlines())))
    print(part1(changes))
    times_sum = 0
    for _ in range(101):
        start = time()
        part2(changes)
        times_sum += time() - start
    print(f'Mean over 100 runs: {(times_sum / 100) * 1000}ms')

File: day01.rb
# frozen_string_literal: true

require 'set'
require 'benchmark'

def part1(changes)
  changes.sum
end

def part2(changes)
  frequencies = Set.new
  value = 0
  loop do
    changes.each do |change|
      value += change
      if frequencies.include? value
        return value
      else
        frequencies.add value
      end
    end
  end
end

lines = File.readlines('inputs/01')
changes = lines.map(&:to_i)

p part1(changes)
p part2(changes)

total_seconds = 0
n = 5000
n.times do
  start = Time.now
  part2(changes)
  total_seconds += (Time.now - start)
end

puts "Mean over #{n} runs: #{(total_seconds / n) * 1000}ms"
